rfd/0175-spiffe-federation.md

```diff
@@ -0,0 +1,429 @@
+---
+author: Noah Stride (noah@goteleport.com)
+state: draft
+---
+
+# RFD 175 - SPIFFE Federation and SPIFFE Join Method
+
+## Required Approvers
+
+- Engineering: @timothyb89
+- Product: @thedevelopnik
+
+## What
+
+Implement support for SPIFFE Federation within Teleport Workload Identity,
+including support for a SPIFFE based join method based on the federation
+mechanism.
+
+See https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.md
+for the SPIFFE Federation specification. This will be referred to as
+"the specification" throughout this document.
+
+This feature will be gated behind a Teleport Enterprise license.
+
+## Why
+
+Implementing support for SPIFFE Federation has the following benefits:
+
+- Improves the feature parity of Workload Identity with competing solutions,
+  increasing appeal to prospects.
+- Allows customers to adopt Teleport Workload Identity gradually, by federating
+  with existing SPIFFE based systems, rather than needing to perform a complete
+  migration.
+- Supports an Active/Active and Active/Passive high availability model.
+- Support segmented environments using multiple Teleport clusters which
+  federate with one another.
+- Enables the implementation of a SPIFFE-based join method.
+
+Implementing support for a SPIFFE join method:
+
+- Secure alternative to the `token` based join-method for OnPrem environments.
+- Allows for more advanced join implementations based on Teleport Workload 
+  Identity, e.g distributing SPIFFE SVIDs using an out of band mechanism and
+  then joining using the SPIFFE SVIDs.
+
+## Implementation
+
+### Federation
+
+SPIFFE federation relationships are one-way. One trust domain can be configured
+to trust identities issued by another trust domain, but the reverse may not 
+be true. However, it is typical that two federation relationships will be
+formed, one in each direction, to allow workloads in both trust domains to
+verify one another.
+
+The federation relies on the exchange of trust bundles. Trust domains which
+wish to support another trust domain trusting identities it has issued must
+publish an accessible Bundle Endpoint. This endpoint provides the trust bundle
+necessary to verify the identities issued by the trust domain.
+
+The structure of this trust bundle is defined in the specification as a JWKS
+(JSON Web Key Set).
+
+#### Teleport's Bundle Endpoint
+
+A new endpoint will be exposed on the Teleport Proxy API, which will be a 
+SPIFFE Federation specification compliant Bundle Endpoint. This will allow
+other SPIFFE trust domains to trust identities issued by Teleport Workload
+Identity.
+
+The endpoint will implement the `https_web` profile, meaning that:
+
+- It will be served using the TLS certificates that the Teleport Proxy has been
+  configured with. These must be certificates issued by a CA trusted by the
+  trust domain control plane which will be consuming the bundle.
```

Comment:
To clarify my own knowledge here, the CA is likely to be something like Let's Encrypt from the certs we automatically retrieve on startup by default, or some other generally trusted CA, but we're covering this in case a company is using a PKI, correct?

Reply:
Right yeah - realistically this is just any Web PKI. This is really just however you've configured TLS to work for your Teleport Proxy already, be that LetsEncrypt or purchased certificates from a CA.